### PR TITLE
tuxtyping: init at 1.8.3

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
I have a 5 year old. So why not.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/games/tuxtyping/default.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  src = fetchurl {
+    url = "https://sourceforge.net/projects/tuxtype/files/tuxtype-source/TuxType%20${version}%20-%20Source/tuxtype_w_fonts-${version}.tar.gz/download";
+    sha256 = "078jvjizycra4b7wp784gc19jnn4jm125939pqip1pwn5rm95c20";
+  };
+
+  patchPhase = ''
+    patchShebangs data/scripts/sed-linux.sh
+  '';
+
+  nativeBuildInputs = [ pkgconfig SDL SDL_image SDL_mixer SDL_ttf librsvg ];
```
Comment (jtojnar) [inline pkgs/games/tuxtyping/default.nix]: Dependencies that are run during build should go to `nativeBuildInputs`, runtime dependencies go to `buildInputs`:

```suggestion
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ SDL SDL_image SDL_mixer SDL_ttf librsvg ];
```

File: pkgs/games/tuxtyping/default.nix [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{ stdenv , fetchurl, pkgconfig, SDL, SDL_image, SDL_mixer, SDL_ttf, librsvg }:
+
+stdenv.mkDerivation rec {
+  version = "1.8.0";
+  name = "tux-type-${version}";
+
+  src = fetchurl {
+    url = "https://sourceforge.net/projects/tuxtype/files/tuxtype-source/TuxType%20${version}%20-%20Source/tuxtype_w_fonts-${version}.tar.gz/download";
```
Comment (jtojnar) [inline pkgs/games/tuxtyping/default.nix]: This seems to be the upstream now https://salsa.debian.org/tux4kids-pkg-team/tuxtype and they have 1.8.3:

https://salsa.debian.org/tux4kids-pkg-team/tuxtype/blob/debian/doc/ChangeLog

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/games/tuxtyping/default.nix [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    patchShebangs data/scripts/sed-linux.sh
+  '';
+
+  nativeBuildInputs = [ pkgconfig SDL SDL_image SDL_mixer SDL_ttf librsvg ];
+
+  configureFlags = [ "--without-sdlpango" ];
+
+  unpackPhase = "tar xf ${src}";
+  sourceRoot = "tuxtype_w_fonts-${version}";
+}
```
Comment (symphorien) [inline pkgs/games/tuxtyping/default.nix]: the `meta` attribute is missing.

File: pkgs/games/tuxtyping/default.nix (hunk 2 above)
Comment (aanderse) [reply pkgs/games/tuxtyping/default.nix]: So I tried switching to 1.8.3 and it added a dependency on SDL_net (what??) and then tried to access a git repository that doesn't exist anymore...

Comment (jtojnar) [reply pkgs/games/tuxtyping/default.nix]: Can you share the expression or the error?

Comment (aanderse) [reply pkgs/games/tuxtyping/default.nix]: Sure:

checking for libxml-2.0... yes
checking for t4k_common... no
configure: t4k_common not located by pkg-config, checking with AC CHECK LIB:
checking for InitT4KCommon in -lt4k_common... no
configure: error: libt4k_common is required: git://git.debian.org/tux4kids/t4kcommon.git

Comment (aanderse) [reply pkgs/games/tuxtyping/default.nix]: I see the repository has moved to https://salsa.debian.org/tux4kids-pkg-team/t4kcommon but doesn't really seem to have anything of value in it.

I'm also very confused about the SDL_net dependency.

Comment (jtojnar) [reply pkgs/games/tuxtyping/default.nix]: I found it here https://github.com/tux4kids/t4kcommon

Comment (aanderse) [reply pkgs/games/tuxtyping/default.nix]: Looks like debian did some weird stuff in some of their branches (several 1.8.3-x releases just to change notes). I grabbed the code from a branch marked as upstream, so we'll go with that.

Comment (jtojnar) [reply pkgs/games/tuxtyping/default.nix]: 1.8.3 is the upstream release, dash releases are just packaging changes.

**Comment by aanderse** [comment]
@jtojnar please let me know if any other changes required.

File: pkgs/games/tuxtype/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, fetchurl, pkgconfig, librsvg, SDL, SDL_image, SDL_mixer, SDL_ttf }:
+
+stdenv.mkDerivation rec {
+  version = "1.8.3";
+  name = "tuxtype-${version}";
+
+  src = fetchurl {
```
Comment (jtojnar) [inline pkgs/games/tuxtype/default.nix]: You should use `fetchFromGitHub` with `rev = "upstream/${version}"`.

File: pkgs/games/tuxtype/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, fetchurl, pkgconfig, librsvg, SDL, SDL_image, SDL_mixer, SDL_ttf }:
+
+stdenv.mkDerivation rec {
+  version = "1.8.3";
+  name = "tuxtype-${version}";
```
Comment (jtojnar) [inline pkgs/games/tuxtype/default.nix]: Nowadays, it is possible to use

```suggestion
  pname = "tuxtype";
```

`name = "${pname}-${version}"` will be created automatically.

File: pkgs/games/tuxtype/default.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  src = fetchurl {
+    url = "https://github.com/tux4kids/tuxtype/archive/upstream/${version}.tar.gz";
+    sha256 = "0cv935ir14cd2c8bgsxxpi6id04f61170gslakmwhxn6r3pbw0lp";
+  };
+
+  patchPhase = ''
```
Comment (jtojnar) [inline pkgs/games/tuxtype/default.nix]: use `postPatch` so that `patches` can still work.

Comment (aanderse) [reply pkgs/games/tuxtype/default.nix]: Thanks

File: pkgs/games/tuxtype/default.nix (hunk 5 above)
Comment (aanderse) [reply pkgs/games/tuxtype/default.nix]: Thanks

File: pkgs/games/tuxtype/default.nix (hunk 4 above)
Comment (aanderse) [reply pkgs/games/tuxtype/default.nix]: Weird... the source code is different when i do this.

  src = fetchFromGitHub {
    owner = "tux4kids";
    repo = "tuxtype";
    rev = "upstream/${version}";
    sha256 = "1i33rhi9gpzfml4hd73s18h6p2s8zcr26va2vwf2pqqd9fhdwpsg";
  };

substitute(): ERROR: file 'Makefile.in' does not exist
builder for '/nix/store/y26mjr7nfg5if9k06k77zai964gd8xmr-tuxtype-1.8.3.drv' failed with exit code 1

Makefile.in existed when I used fetchurl but not when I use fetchFromGitHub.

Comment (jtojnar) [reply pkgs/games/tuxtype/default.nix]: Are you sure that you used the correct hash before? `fetchFromGitHub` will resolve to the same URL.

Comment (jtojnar) [reply pkgs/games/tuxtype/default.nix]: `Makefile.in` would not be in the VCS repository. It is generated by `autoconf` from `Makefile.am` (add `autoreconfHook` to `nativeBuildInputs`).

File: pkgs/games/t4kcommon/default.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, fetchFromGitHub, cmake, pkgconfig, SDL, SDL_image, SDL_mixer, SDL_net, SDL_ttf, libpng, librsvg, libxml2 }:
+
+stdenv.mkDerivation rec {
+  version = "0.1.1";
+  name = "t4kcommon-${version}";
+
+  src = fetchFromGitHub {
+    owner = "tux4kids";
+    repo = "t4kcommon";
+    rev = "ade707815de62d8816723ed6de57cc7955414b5d";
```
Comment (jtojnar) [inline pkgs/games/t4kcommon/default.nix]:
```suggestion
    rev = "upstream/${version}";
```

File: pkgs/games/t4kcommon/default.nix [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, fetchFromGitHub, cmake, pkgconfig, SDL, SDL_image, SDL_mixer, SDL_net, SDL_ttf, libpng, librsvg, libxml2 }:
+
+stdenv.mkDerivation rec {
+  version = "0.1.1";
+  name = "t4kcommon-${version}";
```
Comment (jtojnar) [inline pkgs/games/t4kcommon/default.nix]:
```suggestion
  pname = "t4kcommon";
```

File: pkgs/games/t4kcommon/default.nix [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+  src = fetchFromGitHub {
+    owner = "tux4kids";
+    repo = "t4kcommon";
+    rev = "ade707815de62d8816723ed6de57cc7955414b5d";
+    sha256 = "13q02xpmps9qg8zrzzy2gzv4a6afgi28lxk4z242j780v0gphchp";
+  };
+
```
Comment (jtojnar) [inline pkgs/games/t4kcommon/default.nix]: Can you add `fetchurl` https://salsa.debian.org/tux4kids-pkg-team/t4kcommon/raw/f7073fa384f5a725139f54844e59b57338b69dc7/debian/patches/libpng16.patch to `patches` with a comment that it comes from https://salsa.debian.org/tux4kids-pkg-team/t4kcommon/tree/debian/debian/patches. That might fix the need for old libpng.

**Comment by aanderse** [comment]
@GrahamcOfBorg build tuxtype

File: pkgs/games/t4kcommon/default.nix [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  src = fetchFromGitHub {
+    owner = "tux4kids";
+    repo = "t4kcommon";
+    rev = "upstream/${version}";
+    sha256 = "13q02xpmps9qg8zrzzy2gzv4a6afgi28lxk4z242j780v0gphchp";
+  };
+
+  patches = [
+    # patch from debian to support libpng16 instead of libpng12
+    (fetchpatch {
```
Comment (jtojnar) [inline pkgs/games/t4kcommon/default.nix]: `fetchurl` is sufficient since this is a static file.

File: pkgs/games/t4kcommon/default.nix [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      url = "https://salsa.debian.org/tux4kids-pkg-team/t4kcommon/raw/f7073fa384f5a725139f54844e59b57338b69dc7/debian/patches/libpng16.patch";
+      sha256 = "1lcpkdy5gvxgljg1vkrxych74amq0gramb1snj2831dam48is054";
```
Comment (jtojnar) [inline pkgs/games/t4kcommon/default.nix]:
```suggestion
      sha256 = "10h1sg2sl3p7xhddafkmig68qpr5v7hpv79x7hsv3g5jydb3rr3a";
```

File: pkgs/games/tuxtype/default.nix (hunk 4 above)
Comment (aanderse) [reply pkgs/games/tuxtype/default.nix]: Thanks for all your help on this.

**Comment by aanderse** [comment]
I think that should do it...

**Comment by aanderse** [comment]
@jtojnar merge request to fix configure.ac submitted as requested: https://salsa.debian.org/tux4kids-pkg-team/tuxtype/merge_requests/1

**Comment by jtojnar** [comment]
Looks good, thanks.

**Pull request merged**